File: pm-toolkit/scripts/serve.js
#!/usr/bin/env node

/**
 * PM Toolkit — Single-file preview server
 *
 * Serves panel.html and provides read/write API for ONE Mermaid file.
 *
 *   GET  /              → panel HTML
 *   GET  /api/read      → { source, mtime }
 *   POST /api/write     → body: { source } → saves to file
 *
 * Usage: node serve.js --file <path.mmd> [--port 9876]
 */

const http = require("http");
const fs = require("fs");
const path = require("path");

const args = process.argv.slice(2);
const getArg = (n, fb) => { const i = args.indexOf("--" + n); return i !== -1 && args[i + 1] ? args[i + 1] : fb; };

const FILE = path.resolve(getArg("file", ""));
const PORT = parseInt(getArg("port", "9876"), 10);
const MAX_BODY_BYTES = 1024 * 1024;

if (!FILE) { console.error("用法: node serve.js --file <path.mmd>"); process.exit(1); }
if (!fs.existsSync(FILE)) { console.error("文件不存在: " + FILE); process.exit(1); }

const PANEL = fs.readFileSync(path.join(__dirname, "panel.html"), "utf-8");
const FNAME = path.basename(FILE);

function sendJson(res, statusCode, payload) {
  res.writeHead(statusCode, {
    "Content-Type": "application/json; charset=utf-8",
    "Cache-Control": "no-store",
  });
  res.end(JSON.stringify(payload));
}

function sendText(res, statusCode, text, type = "text/plain; charset=utf-8") {
  res.writeHead(statusCode, {
    "Content-Type": type,
    "Cache-Control": "no-store",
  });
  res.end(text);
}

const server = http.createServer((req, res) => {
  res.setHeader("Access-Control-Allow-Origin", "*");
  res.setHeader("Access-Control-Allow-Methods", "GET,POST,OPTIONS");
  res.setHeader("Access-Control-Allow-Headers", "Content-Type");
  if (req.method === "OPTIONS") { res.writeHead(204); res.end(); return; }

  const url = new URL(req.url, `http://localhost:${PORT}`);

  if (url.pathname === "/" || url.pathname === "/index.html") {
    sendText(res, 200, PANEL, "text/html; charset=utf-8");
    return;
  }

  if (url.pathname === "/api/read") {
    try {
      const source = fs.readFileSync(FILE, "utf-8");
      const mtime = fs.statSync(FILE).mtimeMs;
      sendJson(res, 200, { file: FNAME, source, mtime });
    } catch (e) {
      sendText(res, 500, e.message || "Read failed");
    }
    return;
  }

  if (url.pathname === "/api/write" && req.method === "POST") {
    let body = "", size = 0, tooLarge = false;
    req.on("data", c => {
      if (tooLarge) return;
      size += c.length;
      if (size > MAX_BODY_BYTES) {
        tooLarge = true;
        sendText(res, 413, "Payload too large");
        req.destroy();
        return;
      }
      body += c;
    });
    req.on("end", () => {
      if (tooLarge) return;
      try {
        let parsed;
        try {
          parsed = JSON.parse(body);
        } catch (_) {
          sendText(res, 400, "Invalid JSON");
          return;
        }
        const { source } = parsed;
        if (typeof source !== "string") { sendText(res, 400, "Invalid source"); return; }
        fs.writeFileSync(FILE, source, "utf-8");
        sendJson(res, 200, { ok: true, file: FNAME, mtime: fs.statSync(FILE).mtimeMs });
      } catch (e) {
        sendText(res, 500, e.message || "Write failed");
      }
    });
    req.on("error", (e) => {
      if (!res.writableEnded) sendText(res, 500, e.message || "Request failed");
    });
    return;
  }

  sendText(res, 404, "Not found");
});

server.listen(PORT, () => {
  console.log(`\n🎨 PM Toolkit Preview`);
  console.log(`   文件: ${FILE}`);
  console.log(`   地址: http://localhost:${PORT}`);
  console.log(`   渲染: Mermaid.js (CDN, 浏览器端)\n`);
});

process.on("SIGINT", () => { console.log("\n👋"); server.close(() => process.exit(0)); });
process.on("SIGTERM", () => server.close(() => process.exit(0)));

File: pm-toolkit/scripts/serve.py
#!/usr/bin/env python3
"""PM Toolkit — Single-file preview server

Serves panel.html and provides read/write API for ONE Mermaid file.

  GET  /              → panel HTML
  GET  /api/read      → { source, mtime }
  POST /api/write     → body: { source } → saves to file

Usage: python3 serve.py --file <path.mmd> [--port 9876]
"""

import argparse
import json
import sys
from http.server import HTTPServer, BaseHTTPRequestHandler
from pathlib import Path

MAX_BODY_BYTES = 1024 * 1024
SCRIPT_DIR = Path(__file__).resolve().parent


def parse_args():
    p = argparse.ArgumentParser(description="PM Toolkit Mermaid Preview Server")
    p.add_argument("--file", required=True, help="Path to .mmd file")
    p.add_argument("--port", type=int, default=9876, help="Server port (default: 9876)")
    return p.parse_args()


class PreviewHandler(BaseHTTPRequestHandler):
    panel_html: str = ""
    target_file: Path = Path()
    target_name: str = ""

    def log_message(self, format, *args):
        pass  # silence default logging

    def _cors_headers(self):
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "GET,POST,OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type")
        self.send_header("Cache-Control", "no-store")

    def _send_json(self, code: int, payload: dict):
        body = json.dumps(payload, ensure_ascii=False).encode()
        self.send_response(code)
        self.send_header("Content-Type", "application/json; charset=utf-8")
        self._cors_headers()
        self.end_headers()
        self.wfile.write(body)

    def _send_text(self, code: int, text: str, content_type: str = "text/plain; charset=utf-8"):
        body = text.encode()
        self.send_response(code)
        self.send_header("Content-Type", content_type)
        self._cors_headers()
        self.end_headers()
        self.wfile.write(body)

    def do_OPTIONS(self):
        self.send_response(204)
        self._cors_headers()
        self.end_headers()

    def do_GET(self):
        path = self.path.split("?")[0]

        if path in ("/", "/index.html"):
            self._send_text(200, self.panel_html, "text/html; charset=utf-8")
            return

        if path == "/api/read":
            try:
                source = self.target_file.read_text(encoding="utf-8")
                mtime = self.target_file.stat().st_mtime * 1000
                self._send_json(200, {"file": self.target_name, "source": source, "mtime": mtime})
            except Exception as e:
                self._send_text(500, str(e))
            return

        self._send_text(404, "Not found")

    def do_POST(self):
        path = self.path.split("?")[0]

        if path == "/api/write":
            length = int(self.headers.get("Content-Length", 0))
            if length > MAX_BODY_BYTES:
                self._send_text(413, "Payload too large")
                return
            raw = self.rfile.read(length)
            try:
                data = json.loads(raw)
            except (json.JSONDecodeError, ValueError):
                self._send_text(400, "Invalid JSON")
                return
            source = data.get("source")
            if not isinstance(source, str):
                self._send_text(400, "Invalid source")
                return
            try:
                self.target_file.write_text(source, encoding="utf-8")
                mtime = self.target_file.stat().st_mtime * 1000
                self._send_json(200, {"ok": True, "file": self.target_name, "mtime": mtime})
            except Exception as e:
                self._send_text(500, str(e))
            return

        self._send_text(404, "Not found")


def main():
    args = parse_args()
    target = Path(args.file).resolve()
    if not target.exists():
        print(f"文件不存在: {target}", file=sys.stderr)
        sys.exit(1)

    panel_path = SCRIPT_DIR / "panel.html"
    panel_html = panel_path.read_text(encoding="utf-8") if panel_path.exists() else "<h1>panel.html not found</h1>"

    PreviewHandler.panel_html = panel_html
    PreviewHandler.target_file = target
    PreviewHandler.target_name = target.name

    server = HTTPServer(("", args.port), PreviewHandler)
    print(f"\n🎨 PM Toolkit Preview")
    print(f"   文件: {target}")
    print(f"   地址: http://localhost:{args.port}")
    print(f"   渲染: Mermaid.js (CDN, 浏览器端)\n")

    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print("\n👋")
        server.server_close()


if __name__ == "__main__":
    main()
